presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedTruncatedHistogram.java

```diff
@@ -0,0 +1,318 @@
+    {
+        Block otherValues = other.getValues();
+        LongBigArray otherCounts = other.getCounts();
+        for (int i = 0; i < otherValues.getPositionCount(); i++) {
+            long count = otherCounts.get(i);
+            if (count > 0) {
+                add(i, otherValues, count);
+            }
+        }
+    }
+
+    public void add(int position, Block block, long count)
+    {
+        int hashPosition = getHashPosition(com.facebook.presto.type.TypeUtils.hashPosition(type, block, position), mask);
+
+        // look for an empty slot or a slot containing this key
+        while (true) {
+            if (hashPositions.get(hashPosition) == -1) {
+                break;
+            }
+
+            if (type.equalTo(block, position, values, hashPositions.get(hashPosition))) {
+                counts.add(hashPositions.get(hashPosition), count);
+                return;
+            }
+
+            // increment position and mask to handle wrap around
+            hashPosition = (hashPosition + 1) & mask;
+        }
+
+        addNewGroup(hashPosition, position, block, count);
+    }
+
+    private void addNewGroup(int hashPosition, int position, Block block, long count)
+    {
+        hashPositions.set(hashPosition, values.getPositionCount());
+        counts.set(values.getPositionCount(), count);
+        type.appendTo(block, position, values);
+
+        compactIfNeeded();
+
+        rehashIfNeeded();
+
+        if (getEstimatedSize() > FOUR_MEGABYTES) {
+            throw exceededLocalLimit(new DataSize(4, MEGABYTE));
+        }
+    }
+
+    private void rehashIfNeeded()
+    {
+        // increase capacity, if necessary
+        if (values.getPositionCount() >= maxFill) {
+            int size = maxFill * 2;
+            int hashSize = arraySize(size + 1, FILL_RATIO);
+            mask = hashSize - 1;
+            rehash();
+        }
+    }
+
+    private void rehash()
+    {
+        int hashSize = mask + 1;
+        IntBigArray newHashPositions = new IntBigArray(-1);
+        newHashPositions.ensureCapacity(hashSize);
+
+        for (int i = 0; i < values.getPositionCount(); i++) {
+            // find an empty slot for the address
+            int hashPosition = getHashPosition(com.facebook.presto.type.TypeUtils.hashPosition(type, values, i), mask);
+            while (newHashPositions.get(hashPosition) != -1) {
+                hashPosition = (hashPosition + 1) & mask;
+            }
+
+            // record the mapping
+            newHashPositions.set(hashPosition, i);
+        }
+
+        maxFill = calculateMaxFill(hashSize);
+        hashPositions = newHashPositions;
+
+        this.counts.ensureCapacity(maxFill);
+    }
+
+    /**
+     * Call {@link #compact()} if over {@link #COMPACT_RATIO} of the value/count pairs exceeds {@link #compactLimit}.
+     */
+    private void compactIfNeeded()
+    {
+        if (values.getPositionCount() * COMPACT_RATIO > compactLimit) {
```

Comment:
Yes, it's to keep memory usage low. Checking both memory usage and position count would be more efficient here, and strictly better for fixed-length values. However, for variable-length values like strings, the probability of compaction is proportional to the length, and the value that triggered the compaction is disproportionately often compacted out, so long values might become underrepresented.